nova/virt/vmwareapi/vm_util.py

```diff
@@ -275,20 +274,61 @@ def _vm_ref_cache(id, func, session, data):
 
 
 def vm_ref_cache_from_instance(func):
-    @functools.wraps(func)
+    @six.wraps(func)
     def wrapper(session, instance):
-        id = instance.uuid
-        return _vm_ref_cache(id, func, session, instance)
+        id_ = instance.uuid
+        return _vm_ref_cache(id_, func, session, instance)
     return wrapper
 
 
 def vm_ref_cache_from_name(func):
-    @functools.wraps(func)
+    @six.wraps(func)
     def wrapper(session, name):
-        id = name
-        return _vm_ref_cache(id, func, session, name)
+        id_ = name
+        return _vm_ref_cache(id_, func, session, name)
     return wrapper
 
+
+def vm_ref_cache_heal_from_instance(func):
+    """Decorator for a function working with a cached ManagedObject reference
+    for an instance
+
+    Invalidates the cache in case of a ManagedObjectNotFoundException
+    Most functions rely on the reference to be stable over the life-time of an
+    instance
+    However a vm-migration can move the vm out of the host, and we would end up
+    with a stale cache entry.
+    An operator can also deregister and re-register a vm, or need to recover
+    the vsphere service, all causing then a ManagedObjectNotFoundException
+    at a later point, requiring a restart to clear the cache.
+
+    It requires, that the function in question is idempotent (up to the point
+    where the vm_ref is being used)
+    """
+    @six.wraps(func)
+    def wrapper(session, instance, *args, **kwargs):
+        try:
+            return func(session, instance, *args, **kwargs)
+        except vexc.ManagedObjectNotFoundException as e:
+            with excutils.save_and_reraise_exception() as ctx:
+                id_ = instance.uuid
+                vm_ref = vm_ref_cache_get(id_)
+                obj = e.details.get("obj")
+                # A different moref may be invalid, nothing we can do about it
+                if obj == vm_ref.value:
```

Comment:
Can we return here? i.e.
```
if obj != vm_ref.value
    return
```
then we don't have that much indentation in the next block. I always prefer if-continue and if-return, because I have to keep less context (the if condition) in myhead.

Reply:
Sure, I will change it